### PR TITLE
Split-horizon and Cluster DNS support

### PR DESCRIPTION
**Pull request by bfallonf**
Initial WIP PR for the dns info. 

@abhgupta @smarterclayton a few thoughts/questions I want to ask:
1. I'm presuming architecture/additional_concepts/networking.adoc is the ideal file. But this may change when I get my head around the concept better
2. I'm also presuming that this is a feature that's included in OpenShift by default and there's nothing needed to implement all this.
3. I'm not seeing anything that's defining what this is including. In the email thread @abhgupta gave me, there's Luke, Brenton, Clayton, and Erik giving a few things it _could_ do, but this is a bit rough and doesn't seem to be solid. Is there a more solid explanation of what's happening with this feature?

Any info is helpful.
Thanks.

### CODE REVIEW

**Comment by abhgupta** [comment]
Context/Background: 
Consider the case where a user creates multiple services - a "frontend" service with multiple pods and a "backend" service with one or more pods. In order for the pods for the frontend service to communicate with the "backend" service, environment variables (for service-ip/service-port/username/password/etc) are created and passed along to the "frontend" pods during creation time. If the service is deleted and recreated, it could get a new IP address and would required the "frontend" pods to be recreated in order to pick up the updated environment variables. Additionally, the "backend" service has to be created prior to the creation of any of the "frontend" pods in order to ensure that service IP is generated and the "frontend" pods can pick it up as an environment variable.

What this change IS:
This creates a DNS for services. With this change, services can be reached via service DNS in addition to the service ip/port. The service DNS format is well defined and has been covered in the email thread. The DNS also allows for wildcard lookups and so, a lookup like xyz.<service-name>.<namespace-name>.local would resolve to the service IP.

What this change DOES/ALLOWS:
This eliminates the need to recreate pods when services are deleted and recreated, resulting in a new IP being created for the service. Pods could use the service DNS instead of relying on environment variables for IP/Port since the DNS would not change. Also, this removes the need to create the "backend" service before any of the "frontend" pods are created since the service name (and hence dns) can be well-known/established upfront.

**Comment by abhgupta** [comment]
A more advanced case is with headless services. These are services for which the kube-proxy does not have any information and does not perform any routing. Service DNS can still be used and it will respond will multiple A records (one for each pod of the service). In this manner DNS can be used to route/load-balance for headless services.

**Comment by bfallonf** [comment]
@abhgupta Thanks so much for that extra information. I think I have a handle on it all now. One more question: You mention "service IP" a lot in the comments above, are you talking about an environment variable? For example: "as any lookups would resolve to the service IP". It may not mean this in every case, but I'm worried I've confused that and the actual IP of the service. (Hoping this question makes sense...). Also, if there's any sentences where I've confused the meaning from your suggestions, please let me know. 

Thanks!

File: architecture/additional_concepts/networking.adoc [hunk 1]
```diff
@@ -22,6 +22,34 @@ recommend that you create a
...
+
+----
+<pod_namespace>.local
+<service>.<pod_namespace>.local
+----
+
+This prevents having to restart frontend pods in order to pick up new services, which results in a new IP being created for the service. This also removes the need to use environment variables, as pods can use the service DNS instead of relying on environment variables since the DNS would not change and you can reference database services as "db.local" in config files. This also supports wildcard lookups, as any lookups would resolve to the service IP, and removes the need to create the backend service before any of the frontend pods, since the service name (and hence DNS) is established upfront.
+
+This DNS structure also covers headless services; where a portal IP is not assigned to the service and the kube-proxy does not load-balance and provide routing for its endpoints. Service DNS can still be used and it will respond will multiple A records, one for each pod of the service, allowing the client to round robin between each pod.
```
Comment (abhgupta) [inline architecture/additional_concepts/networking.adoc]: Would change to --> does not load-balance ~~and~~ **or** provide routing for

File: architecture/additional_concepts/networking.adoc [hunk 2]
```diff
@@ -22,6 +22,34 @@ recommend that you create a
 link:../core_objects/kubernetes_model.html#service[service], then interact with
 the service.
 
+== OpenShift DNS
+
+If you are running multiple services, such as frontend and backend services for use with multiple pods, in order for the frontend pods to communicate with the backend services, environment variables are created for such values as username, service-ip, etc. If the service is deleted and recreated, the IP address is assigned to a new environment variable, and requires the frontend pods to be recreated in order to pick up the updated values. Additionally, the backend service has to be created before any of the frontend pods to ensure that the service IP environment variable is generated properly and that the frontend pods can resolve to it.
```
Comment (abhgupta) [inline architecture/additional_concepts/networking.adoc]: If the service is deleted and recreated, ~~the IP address is assigned to a new environment variable,~~ **a new IP address can be assigned to the service,** and requires the frontend pods to be recreated in order to pick up the updated values **for the service IP environment variable**. Additionally, the backend service has to be created before any of the frontend pods to ensure that the service IP ~~environment variable~~ is generated properly and that the **same can be provide to the frontend pods as an environment variable** ~~frontend pods can resolve to it~~ .

**Comment by abhgupta** [comment]
An IP address can either be specified for a service (by the user) or one is generated/allocated/assigned to the service. An environment variable (for the service IP) is created for pods that are within the same namespace (as the service) so that they can access the service.

**Comment by bfallonf** [comment]
@abhgupta Sure, thanks much for that. I think I've got a handle on it, and I've taken in your suggested edits above. Please let me know if there's anything else you'd like to comment on for this US.

Also, @adellape @CowboysFan @tpoitras @tnguyen-rh this is up for peer review if anyone has any suggestions. A lot of this is a bit wordy so review welcome.

File: architecture/additional_concepts/networking.adoc [hunk 3]
```diff
@@ -22,6 +22,63 @@ recommend that you create a
...
+If you are running multiple
+link:../core_objects/kubernetes_model.html#service[services], such as frontend
+and backend services for use with multiple pods, in order for the frontend pods
+to communicate with the backend services, environment variables are created for
+username, service-ip, etc. If the service is deleted and recreated, a new IP
+address can be assigned to the service, and requires the frontend pods to be
+recreated in order to pick up the updated values for the service IP environment
+variable. Additionally, the backend service has to be created before any of the
+frontend pods to ensure that the service IP is generated properly and that it
+can be provide to the frontend pods as an environment variable.
```
Comment (adellape) [inline architecture/additional_concepts/networking.adoc]: s/be provide/be provided/

File: architecture/additional_concepts/networking.adoc [hunk 4]
```diff
@@ -22,6 +22,63 @@ recommend that you create a
...
+
+For this reason, OpenShift has an inbuilt DNS so that the services can be
```
Comment (adellape) [inline architecture/additional_concepts/networking.adoc]: Maybe "built-in"?

File: architecture/additional_concepts/networking.adoc [hunk 5]
```diff
@@ -22,6 +22,63 @@ recommend that you create a
...
+reached by the service DNS as well as the service IP/port. OpenShift supports
+split DNS by running a SkyDNS on the master that answers DNS queries for
```
Comment (adellape) [inline architecture/additional_concepts/networking.adoc]: Maybe:

s/a SkyDNS/SkyDNS/

And possibly embed link to https://github.com/skynetservices/skydns.

File: architecture/additional_concepts/networking.adoc [hunk 6]
```diff
@@ -22,6 +22,63 @@ recommend that you create a
...
+
+----
+<pod_namespace>.local
+<service>.<pod_namespace>.local
+----
+
+This prevents having to restart frontend pods in order to pick up new services,
+which creates a new IP for the service. This also removes the need to use
+environment variables, as pods can use the service DNS instead of relying on
+the. Also, as the DNS does not change, you can reference database services as
```
Comment (adellape) [inline architecture/additional_concepts/networking.adoc]: "relying on the." ?

File: architecture/additional_concepts/networking.adoc [hunk 7]
```diff
@@ -22,6 +22,63 @@ recommend that you create a
...
+"db.local" in config files. Wildcard lookups are also supported, as any lookups
+resolve to the service IP, and removes the need to create the backend service
+before any of the frontend pods, since the service name (and hence DNS) is
+established upfront.
+
+This DNS structure also covers headless services; where a portal IP is not
```
Comment (adellape) [inline architecture/additional_concepts/networking.adoc]: Seems like strange semi-colon usage, but maybe it's just me.

**Comment by adellape** [comment]
@bfallonf Some minor comments inline. And I figure you're waiting til the end, but, a reminder to squash the commits into one when it's ready for merging.

**Comment by bfallonf** [comment]
@adellape Thanks for the review. Suggestions have been edited in. Ya, was waiting for it all to be over before a squash. Should be good now.

**Comment by smarterclayton** [comment]
I hate to do this but there are changes coming to DNS.  Be aware we'll probably need to update this in a few weeks.  Changes are mostly positional.

**Comment by bfallonf** [comment]
@smarterclayton Blah, ok. Thanks for the heads up.. As long as there's a trello card or something for us to track the docs work and find the info, it should be ok.
